File: lib/qa_bbheader_bb.cc
/* -*- c++ -*- */
/*
 * Copyright 2023 AsriFox.
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include <gnuradio/analog/random_uniform_source.h>
#include <gnuradio/attributes.h>
#include <gnuradio/blocks/head.h>
#include <gnuradio/blocks/tag_debug.h>
#include <gnuradio/dvbs2tx/bbheader_bb.h>
#include <gnuradio/dvbs2tx/dvbs2_config.h>
#include <gnuradio/top_block.h>
#include <boost/test/unit_test.hpp>

namespace gr {
namespace dvbs2tx {

using namespace gr::dtv;

BOOST_AUTO_TEST_CASE(test_bbheader_debug_tags)
{
    auto tb = gr::make_top_block("test_bbheader");
    auto src = gr::analog::random_uniform_source_b::make(0, 255, 0);
    auto head = gr::blocks::head::make(sizeof(u8), 64800);
    auto core = bbheader_bb::make(FECFRAME_SHORT, C8_9, MOD_16APSK, false, 0, RO_0_20);
    auto snk = gr::blocks::tag_debug::make(sizeof(u8), "bbheader", "modcod");
    tb->connect(src, 0, head, 0);
    tb->connect(head, 0, core, 0);
    tb->connect(core, 0, snk, 0);
    tb->run();
    tb->wait();
}

} /* namespace dvbs2tx */
} /* namespace gr */

File: python/dvbs2tx/qa_bbheader_bb.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2023 AsriFox.
#
# SPDX-License-Identifier: GPL-3.0-or-later
#

from gnuradio import gr, gr_unittest
from gnuradio import analog, blocks, dtv
try:
    from gnuradio.dvbs2tx import bbheader_bb
except ImportError:
    import os
    import sys
    dirname, filename = os.path.split(os.path.abspath(__file__))
    sys.path.append(os.path.join(dirname, "bindings"))
    from gnuradio.dvbs2tx import bbheader_bb
from time import sleep

class qa_bbheader_bb(gr_unittest.TestCase):

    def setUp(self):
        self.tb = gr.top_block()
        self.src = analog.random_uniform_source_b(0, 255, 0)
        self.snk = blocks.tag_debug(gr.sizeof_char, 'bbheader', 'modcod')

    def tearDown(self):
        self.tb.disconnect_all()
        self.tb = None
        self.src = None
        self.snk = None

    def test_instance(self):
        instance = bbheader_bb(
            dtv.FECFRAME_NORMAL, 
            dtv.C1_4, 
            dtv.MOD_QPSK,
            True, 
            0, 
            dtv.RO_0_35)

    def test_print_tags(self):
        instance = bbheader_bb(
            dtv.FECFRAME_SHORT,
            dtv.C8_9,
            dtv.MOD_16APSK,
            False,
            0,
            dtv.RO_0_20)
        self.tb.disconnect_all()
        self.tb.connect(self.src, instance, self.snk)
        self.tb.run()
        sleep(3)
        self.tb.stop()
        self.tb.wait()
        

if __name__ == '__main__':
    gr_unittest.run(qa_bbheader_bb)
